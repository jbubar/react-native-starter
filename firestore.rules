rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /todoLists/{listID} {
      allow read, create, update, delete: if true;

      match /todos/{todoID} {
        function notEmptyTodo() {
          return request.resource.data.description.size() > 0;
        }

        function isNewTodo() {
          return request.resource.data.done == false;
        }

        allow list: if true;
        allow create: if notEmptyTodo() && isNewTodo();
        allow update: if notEmptyTodo();
        allow delete: if true;
      }
    }
  }
}