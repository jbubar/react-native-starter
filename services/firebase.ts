import firebase from 'firebase/app';
import 'firebase/firestore';
import {Platform} from 'react-native';

const config = {
  apiKey: 'AIzaSyCMoApbCmvr3_uL2yGZoZ7UcU24h0rhlgU',
  authDomain: 'interview-todos.firebaseapp.com',
  projectId: 'interview-todos',
  storageBucket: 'interview-todos.appspot.com',
  messagingSenderId: '598035300081',
  appId: '1:598035300081:web:dc63e598fe7998b2f5f350',
  measurementId: 'G-QY9P3TP4PH',
};

console.log(config);

firebase.initializeApp(config);

// Experimental Long Polling Open Issue - https://github.com/firebase/firebase-js-sdk/issues/1674
firebase
  .firestore()
  .settings({experimentalForceLongPolling: Platform.OS === 'android'});
