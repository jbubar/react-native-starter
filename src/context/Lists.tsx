import React, {createContext, useState, useCallback, useEffect} from 'react';
import firebase from 'firebase/app';

import {useCollectionData} from 'react-firebase-hooks/firestore';

export type Todo = {
  id: string;
  name: string;
};

const defaultState = {
  lists: [],
  selectedList: null,
};

const defaultActions = {
  addList: async (_newTodo: string) => {},
  selectList: (_todo: Todo) => {},
};

export const ListContext = createContext({
  isLoading: false,
  ...defaultState,
  ...defaultActions,
});

const ListProvider: React.FC = ({children}) => {
  const [selectedList, selectList] = useState({});

  const ListsCollection = firebase.firestore().collection('todoLists');

  const [lists, isLoading, error] = useCollectionData(ListsCollection, {
    idField: 'id',
  });
  useEffect(() => {
    if (lists) {
      selectList(lists[0]);
    }
  }, [lists]);

  const addList = useCallback(
    async ({name}) => {
      if (name.length) {
        try {
          return await ListsCollection.add({
            name,
          });
        } catch ({message}) {
          console.error(message);
        }
      }
    },
    [ListsCollection],
  );

  return (
    <ListContext.Provider
      value={{
        lists,
        isLoading,
        error: error?.message,
        addList,
        selectedList,
        selectList,
      }}
      children={children}
    />
  );
};

export const MockedProvider: React.FC = props => (
  <ListContext.Provider {...props} />
);

export default ListProvider;
