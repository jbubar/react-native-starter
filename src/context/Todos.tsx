import React, {
  createContext,
  useState,
  useCallback,
  useContext,
  useMemo,
} from 'react';
import firebase from 'firebase/app';

import {useCollectionData} from 'react-firebase-hooks/firestore';
import {ListContext} from './Lists';

export type Todo = {
  id: string;
  description: string;
  done: boolean;
  date?: Date;
};

const defaultState = {
  todos: [],
  selectedTodo: null,
};

const defaultActions = {
  addTodo: async (_newTodo: string) => {},
  toggleDone: async (_todo: Todo) => {},
  saveTodo: async (_todo: Todo) => {},
  deleteTodo: async (_todo: Todo) => {},
  setSelected: (_todo: Todo) => {},
};

export const TodoContext = createContext({
  isLoading: false,
  ...defaultState,
  ...defaultActions,
});

const TodoProvider: React.FC = ({children}) => {
  const {selectedList} = useContext(ListContext);

  const [selectedTodo, setSelected] = useState(null);

  const TodosCollection = useMemo(
    () =>
      firebase
        .firestore()
        .collection('todoLists')
        .doc(selectedList.id)
        .collection('todos'),
    [selectedList],
  );

  const [todos, isLoading, error] = useCollectionData(TodosCollection, {
    idField: 'id',
  });

  const addTodo = useCallback(
    async newTodo => {
      if (newTodo.length) {
        try {
          await TodosCollection.add({
            description: newTodo,
            done: false,
          });
        } catch ({message}) {
          console.error(message);
        }
      }
    },
    [TodosCollection],
  );

  const toggleDone = useCallback(
    async ({id, done}: Todo) => {
      try {
        await TodosCollection.doc(id).update({done: !done});
      } catch (e) {
        console.error(e);
      }
    },
    [TodosCollection],
  );

  const saveTodo = useCallback(
    async ({id, description, done, date}: Todo) => {
      try {
        await TodosCollection.doc(id).update({
          description,
          done,
          date: firebase.firestore.Timestamp.fromDate(date),
        });
      } catch (e) {
        console.error(e);
      }
    },
    [TodosCollection],
  );

  const deleteTodo = useCallback(
    async ({id}: Todo) => {
      try {
        await TodosCollection.doc(id).delete();
      } catch (e) {
        console.error(e);
      }
    },
    [TodosCollection],
  );

  return (
    <TodoContext.Provider
      value={{
        todos,
        isLoading,
        error: error?.message,
        addTodo,
        toggleDone,
        selectedTodo,
        setSelected,
        saveTodo,
        deleteTodo,
      }}
      children={children}
    />
  );
};

export const MockedProvider: React.FC = props => (
  <TodoContext.Provider {...props} />
);

export default TodoProvider;
