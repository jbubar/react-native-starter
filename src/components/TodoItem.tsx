import React, {useContext} from 'react';
import {StyleSheet, TouchableOpacity} from 'react-native';

import {formatDistance, compareAsc} from 'date-fns';
import {ListItem, Icon, Text, Layout} from '@ui-kitten/components';

import {Todo, TodoContext} from '../context/Todos';

const styles = StyleSheet.create({
  icon: {margin: 0},
  strikethrough: {
    textDecorationLine: 'line-through',
    textDecorationStyle: 'solid',
  },
});

export default ({item}: {item: Todo}) => {
  const {id, done, date, description} = item;
  const {toggleDone, setSelected} = useContext(TodoContext);

  return (
    <ListItem
      onPress={() => setSelected(item)}
      testID="todoItem"
      title={() => (
        <Text key={id} style={[done && styles.strikethrough]}>
          {description}
        </Text>
      )}
      description={() =>
        date ? (
          <Layout style={{flexDirection: 'row'}}>
            <Text style={[done && styles.strikethrough, {marginRight: 10}]}>
              {formatDistance(date.toDate(), new Date(), {addSuffix: true})}
            </Text>
            <Text style={[done && styles.strikethrough, {color: 'red'}]}>
              {compareAsc(date.toDate(), new Date()) < 1 ? 'OVERDUE' : ''}
            </Text>
          </Layout>
        ) : (
          <></>
        )
      }
      accessoryRight={props => (
        <TouchableOpacity onPress={() => toggleDone(item)} testID="toggleDone">
          <Icon
            name={done ? 'checkmark-square-outline' : 'square-outline'}
            style={[props?.style, styles.icon]}
          />
        </TouchableOpacity>
      )}
    />
  );
};
