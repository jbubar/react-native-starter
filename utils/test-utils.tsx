// test-utils.js

import React, {ReactElement} from 'react';
import {SafeAreaView, StyleSheet} from 'react-native';

import * as eva from '@eva-design/eva';
import {EvaIconsPack} from '@ui-kitten/eva-icons';
import {ApplicationProvider, IconRegistry} from '@ui-kitten/components';

import {render, RenderOptions} from '@testing-library/react-native';

const styles = StyleSheet.create({
  safeArea: {flex: 1},
});

const AllTheProviders: React.FC = ({children}) => {
  return (
    <>
      <IconRegistry icons={EvaIconsPack} />
      <ApplicationProvider {...eva} theme={eva.light}>
        <SafeAreaView style={styles.safeArea}>{children}</SafeAreaView>
      </ApplicationProvider>
    </>
  );
};

const customRender = (ui: ReactElement, options: RenderOptions) =>
  render(ui, {wrapper: AllTheProviders, ...options});

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
