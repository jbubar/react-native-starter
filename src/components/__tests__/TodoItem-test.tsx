/* eslint-disable @typescript-eslint/no-unused-vars */
import React from 'react';
import {render, fireEvent} from 'test-utils';

import 'firebase/firestore';

import TodoItem from '../TodoItem';

import {MockedProvider} from '../../context/Todos';
import {waitFor} from '@testing-library/react-native';

jest.mock('react-native-config', () => ({
  Config: {
    SINGLE_LIST_ID: '1',
  },
}));

describe('TodoItem', () => {
  let wrapper, toggleDone;

  const INCOMPLETE_TODO = {
    id: '1',
    description: 'Sample',
    done: false,
  };

  const COMPLETE_TODO = {
    id: '1',
    description: 'Sample',
    done: true,
  };

  beforeAll(() => {
    jest.resetAllMocks();
  });

  beforeEach(() => {
    toggleDone = jest.fn(todo => Promise.resolve());
    wrapper = render(
      <MockedProvider value={{toggleDone}}>
        <TodoItem item={INCOMPLETE_TODO} />
      </MockedProvider>,
    );
  });

  test('Snapshot - Incomplete', () => {
    const {toJSON} = wrapper;
    expect(toJSON()).toMatchSnapshot();
  });

  test('Snapshot - Done', () => {
    const {toJSON} = render(
      <MockedProvider value={{toggleDone}}>
        <TodoItem item={COMPLETE_TODO} />
      </MockedProvider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  test('Toggle Done', async () => {
    const {getByTestId, toJSON} = wrapper;

    const toggleButton = getByTestId('toggleDone');
    fireEvent.press(toggleButton);

    await waitFor(() => {
      expect(toggleDone).toHaveBeenCalledWith(INCOMPLETE_TODO);
    });
  });
});
