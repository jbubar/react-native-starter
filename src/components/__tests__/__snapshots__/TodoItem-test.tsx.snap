// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem Snapshot - Done 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "flexDirection": "row",
          "opacity": 1,
          "paddingHorizontal": 8,
          "paddingVertical": 12,
        }
      }
      testID="todoItem"
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          appearance="default"
          category="p1"
          style={
            Array [
              Object {
                "color": "#222B45",
                "fontFamily": "System",
                "fontSize": 15,
                "fontWeight": "400",
              },
              Array [
                Object {
                  "textDecorationLine": "line-through",
                  "textDecorationStyle": "solid",
                },
              ],
            ]
          }
        >
          Sample
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="toggleDone"
      >
        <View
          style={
            Object {
              "transform": Array [
                Object {
                  "scale": 1,
                },
              ],
            }
          }
        >
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight={24}
            bbWidth={24}
            fill="#8F9BB3"
            focusable={false}
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Array [
                  Object {
                    "height": 24,
                    "marginHorizontal": 8,
                    "tintColor": "#8F9BB3",
                    "width": 24,
                  },
                  Object {
                    "margin": 0,
                  },
                ],
                Object {
                  "flex": 0,
                  "height": 24,
                  "width": 24,
                },
              ]
            }
            vbHeight={24}
            vbWidth={24}
          >
            <RNSVGGroup
              fill={4287601587}
              propList={
                Array [
                  "fill",
                ]
              }
            >
              <RNSVGGroup>
                <RNSVGGroup>
                  <RNSVGRect
                    height="24"
                    opacity={0}
                    width="24"
                    x={0}
                    y={0}
                  />
                  <RNSVGPath
                    d="M20 11.83a1 1 0 0 0-1 1v5.57a.6.6 0 0 1-.6.6H5.6a.6.6 0 0 1-.6-.6V5.6a.6.6 0 0 1 .6-.6h9.57a1 1 0 1 0 0-2H5.6A2.61 2.61 0 0 0 3 5.6v12.8A2.61 2.61 0 0 0 5.6 21h12.8a2.61 2.61 0 0 0 2.6-2.6v-5.57a1 1 0 0 0-1-1z"
                  />
                  <RNSVGPath
                    d="M10.72 11a1 1 0 0 0-1.44 1.38l2.22 2.33a1 1 0 0 0 .72.31 1 1 0 0 0 .72-.3l6.78-7a1 1 0 1 0-1.44-1.4l-6.05 6.26z"
                  />
                </RNSVGGroup>
              </RNSVGGroup>
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;

exports[`TodoItem Snapshot - Incomplete 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "flexDirection": "row",
          "opacity": 1,
          "paddingHorizontal": 8,
          "paddingVertical": 12,
        }
      }
      testID="todoItem"
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          appearance="default"
          category="p1"
          style={
            Array [
              Object {
                "color": "#222B45",
                "fontFamily": "System",
                "fontSize": 15,
                "fontWeight": "400",
              },
              Array [
                false,
              ],
            ]
          }
        >
          Sample
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="toggleDone"
      >
        <View
          style={
            Object {
              "transform": Array [
                Object {
                  "scale": 1,
                },
              ],
            }
          }
        >
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight={24}
            bbWidth={24}
            fill="#8F9BB3"
            focusable={false}
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Array [
                  Object {
                    "height": 24,
                    "marginHorizontal": 8,
                    "tintColor": "#8F9BB3",
                    "width": 24,
                  },
                  Object {
                    "margin": 0,
                  },
                ],
                Object {
                  "flex": 0,
                  "height": 24,
                  "width": 24,
                },
              ]
            }
            vbHeight={24}
            vbWidth={24}
          >
            <RNSVGGroup
              fill={4287601587}
              propList={
                Array [
                  "fill",
                ]
              }
            >
              <RNSVGGroup>
                <RNSVGGroup>
                  <RNSVGRect
                    height="24"
                    opacity={0}
                    width="24"
                    x={0}
                    y={0}
                  />
                  <RNSVGPath
                    d="M18 21H6a3 3 0 0 1-3-3V6a3 3 0 0 1 3-3h12a3 3 0 0 1 3 3v12a3 3 0 0 1-3 3zM6 5a1 1 0 0 0-1 1v12a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1V6a1 1 0 0 0-1-1z"
                  />
                </RNSVGGroup>
              </RNSVGGroup>
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
